WIP / RFC New Reticule (imgui)

ecraven opened the pull request:
This is one way the new reticule can go. Please test and tell us about problems with it.

I will clean up history before this is merged.

See http://pioneerwiki.com/wiki/Flight_UI for some details

![one](https://cloud.githubusercontent.com/assets/215510/20231425/092c28c8-a862-11e6-9c7d-218d3bd043d3.jpg)
![two](https://cloud.githubusercontent.com/assets/215510/20231443/376413c2-a862-11e6-9bf3-46ebf7c4c537.jpg)

impaktor commented:
Sweet! 
Earlier versions had a lot of different bars, or did I imagine that? This new version is no doubt cleaner. Is there any information in those bars I imagined that aught to be shown?

ecraven commented:
> Earlier versions had a lot of different bars, or did I imagine that?

There were bars down at the navball. Those will be added with the navball PR. The braking bar (right side of the reticule) is only shown when actually approaching the target.

impaktor commented on src/LuaBody.cpp:
```diff
@@ -394,6 +438,29 @@ static int l_body_find_nearest_to(lua_State *l)
+		Body *b = LuaObject<Body>::CheckFromLua(1);
+		LuaPush(l, b->GetPhysRadius());
+		return 1;
+}
```
Double indentation?

(Yeah, I know, I only look at the pattern of the text, not actual code :frowning: )

impaktor commented on src/LuaPlayer.cpp:
```diff
@@ -208,6 +208,102 @@ static int l_set_hyperspace_target(lua_State *l)
+}
+	
+static int l_get_is_mouse_active(lua_State *l)
+{
```
Also here is twice the indentation in the github editor,

impaktor commented on src/PiGui.cpp:
```diff
@@ -185,8 +190,8 @@ int PiGui::RadialPopupSelectMenu(const ImVec2& center, std::string popup_id, std
-									 center.x + cosf((item_inner_ang_min + item_inner_ang_max) * 0.5f) * (RADIUS_MIN + RADIUS_MAX) * 0.5f - text_size.x * 0.5f,
-									 center.y + sinf((item_inner_ang_min + item_inner_ang_max) * 0.5f) * (RADIUS_MIN + RADIUS_MAX) * 0.5f - text_size.y * 0.5f);
+															 center.x + cosf((item_inner_ang_min + item_inner_ang_max) * 0.5f) * (RADIUS_MIN + RADIUS_MAX) * 0.5f - text_size.x * 0.5f,
+															 center.y + sinf((item_inner_ang_min + item_inner_ang_max) * 0.5f) * (RADIUS_MIN + RADIUS_MAX) * 0.5f - text_size.y * 0.5f);
```
Here, the indentation looks goofy

impaktor commented on src/PiGui.cpp:
```diff
@@ -220,19 +225,27 @@ bool PiGui::CircularSlider(const ImVec2 &center, float *v, float v_min, float v_
+	const vector2f dataSize(width, height);
+	const Graphics::TextureDescriptor texDesc(Graphics::TEXTURE_RGBA_8888,
+																						dataSize, texSize, Graphics::LINEAR_CLAMP,
+																						false, false, false, 0, Graphics::TEXTURE_2D);
```
...and here? I prefer alidning at the opening ´(´, or one indentation level from the ´cons´

impaktor commented on src/WorldView.cpp:
```diff
@@ -123,8 +123,8 @@ void WorldView::InitObject()
-					formatarg("power", 100.0f * LOW_THRUST_LEVELS[i], "f.0")));
+																			 stringf(Lang::SET_LOW_THRUST_POWER_LEVEL_TO_X_PERCENT,
+																							 formatarg("power", 100.0f * LOW_THRUST_LEVELS[i], "f.0")));
 		m_lowThrustPowerOptions->Add(label, 50, float(ypos));
```
...same.

impaktor commented on src/WorldView.cpp:
```diff
@@ -508,7 +469,7 @@ void WorldView::ResetHyperspaceButton()
 {
 	PROFILE_SCOPED()
-	assert(m_game);
+		assert(m_game);
```
poop

impaktor commented on src/WorldView.cpp:
```diff
@@ -819,44 +775,16 @@ void WorldView::RefreshButtonStateAndVisibility()
+																																						formatarg("system", dest.IsBodyPath() ? s->GetBodyByPath(dest)->GetName() : s->GetName()),
+																																						formatarg("x", dest.sectorX),
+																																						formatarg("y", dest.sectorY),
+																																						formatarg("z", dest.sectorZ)));
```
"Grandma, what long lines you have?"

impaktor commented on src/WorldView.cpp:
```diff
@@ -819,44 +775,16 @@ void WorldView::RefreshButtonStateAndVisibility()
 
 		m_game->GetCpan()->SetOverlayText(ShipCpanel::OVERLAY_TOP_RIGHT, stringf(Lang::JUMP_COMPLETE,
-			formatarg("percent", m_game->GetHyperspaceArrivalProbability()*100.0, "f3.1")));
+																																						 formatarg("percent", m_game->GetHyperspaceArrivalProbability()*100.0, "f3.1")));
```
"So I'll huff and I'll puff"

impaktor commented on src/WorldView.cpp:
```diff
@@ -1058,7 +973,7 @@ void WorldView::RefreshButtonStateAndVisibility()
 				text += "\n";
 				text += stringf(Lang::SHIELD_STRENGTH_N, formatarg("shields",
-					(sShields*0.01f) * float(prop_var))); // At that point, it still holds the property for the shields
+																													 (sShields*0.01f) * float(prop_var))); // At that point, it still holds the property for the shields
```
..."And I'll blow this house down"

nozmajner commented:
A little thing I noticed: If you unset your Nav or Combat target (clicking on it again), the right side data display disappears instead of reverting to the Frame data, and the Switch icons disappear too.
Which is a neat de-clutter mode, but could be tedious for the player to hunt for a target just to be able to switch it back.
Maybe there could be a none button too, which would only show the very basic Frame data on the left side?

nozmajner commented:
A bit of an update to the icons (hooked them up in the HUD lua too), and to the brake-bar overshoot coloring.
I've made the vector markers and crosshairs thinner, so they won't obstruct that much. It was a bit distracting.
https://www.dropbox.com/s/ma9q31u9mfaf4f8/HUD.ZIP?dl=0

![screenshot-20161113-171440](https://cloud.githubusercontent.com/assets/4182678/20247080/c6ae048a-a9c4-11e6-8b4d-f18b89827fdb.png)

walterar commented:
ver  on: Linux

System Name: Linux
Host Name: oper-desktop
Release(Kernel) Version: 4.4.0-45-generic
Kernel Build Timestamp: #66~14.04.1-Ubuntu SMP Wed Oct 19 15:05:38 UTC 2016
Machine Arch: x86_64
Domain Name: (none)

SDL Version 2.0.3
Initialized OpenGL 3.1, with extensions, renderer
invalid axis binding '' in config file for BindAxisPitch
invalid axis binding '' in config file for BindAxisRoll
invalid axis binding '' in config file for BindAxisYaw
started 4 worker threads
ShipType::Init()
Lua::Init()
Server agent disabled
Game::CanLoadGame('_exit')
nanosvg: /media/auxiliar/Building/APIONEER64/pioneer-imgui/data/icons/icons.svg 1024x1024
texture id: 20
nanosvg: /media/auxiliar/Building/APIONEER64/pioneer-imgui/data/icons/icons.svg 1024x1024
texture id: 21
GalaxyGenerator::Init()
Creating new galaxy generator 'legacy' version 1
Number of factions added: 103
StarSystemCache: misses: 0, slave hits: 0, master hits: 0
SectorCache: misses: 101, slave hits: 0, master hits: 2
FaceParts::Init()
Face Generation source images loaded.
new ModelCache
Shields::Init
BaseSphere::Init
CityOnPlanet::Init
CreateCollisionMesh for : (kbuilding01)
CreateCollisionMesh for : (kbuilding02)
CreateCollisionMesh for : (kbuilding03)
CreateCollisionMesh for : (newbuilding1)
CreateCollisionMesh for : (newbuilding10)
CreateCollisionMesh for : (newbuilding11)
CreateCollisionMesh for : (newbuilding2)
CreateCollisionMesh for : (newbuilding3)
CreateCollisionMesh for : (newbuilding4)
CreateCollisionMesh for : (newbuilding5)
CreateCollisionMesh for : (newbuilding6)
CreateCollisionMesh for : (newbuilding7)
CreateCollisionMesh for : (newbuilding8)
CreateCollisionMesh for : (newbuilding9)
Got 14 buildings of tag city_building
 - kbuilding01: 29.664692
 - kbuilding02: 46.268344
 - kbuilding03: 41.542333
 - newbuilding1: 36.202756
 - newbuilding10: 16.297418
 - newbuilding11: 34.961034
 - newbuilding2: 22.822782
 - newbuilding3: 22.649894
 - newbuilding4: 25.978193
 - newbuilding5: 22.908506
 - newbuilding6: 29.817877
 - newbuilding7: 22.674426
 - newbuilding8: 23.154979
 - newbuilding9: 36.528138
End of buildings.
SpaceStation::Init
CreateCollisionMesh for : (ground_station)
ground_station has:
 14 entrances,
 14 pads,
 0 exits
CreateCollisionMesh for : (new_ground)
new_ground has:
 6 entrances,
 6 pads,
 0 exits
CreateCollisionMesh for : (orbital_station_2-10k)
orbital_station_2-10k has:
 5 entrances,
 64 pads,
 5 exits
CreateCollisionMesh for : (orbital_station_2-2k)
orbital_station_2-2k has:
 4 entrances,
 18 pads,
 4 exits
CreateCollisionMesh for : (orbital_station_2-5k)
orbital_station_2-5k has:
 5 entrances,
 64 pads,
 5 exits
CreateCollisionMesh for : (orbital_station_2-5k10k)
orbital_station_2-5k10k has:
 5 entrances,
 64 pads,
 5 exits
NavLights::Init
Sfx::Init
Sound::Init


Loading took: 48531.046247 milliseconds
CreateCollisionMesh for : (ac33)
CreateCollisionMesh for : (amphiesma)
CreateCollisionMesh for : (bluenose)
CreateCollisionMesh for : (bowfin)
CreateCollisionMesh for : (deneb)
CreateCollisionMesh for : (dsminer)
CreateCollisionMesh for : (kanara_civ)
CreateCollisionMesh for : (lunarshuttle)
CreateCollisionMesh for : (malabar)
CreateCollisionMesh for : (molamola)
CreateCollisionMesh for : (natrix)
CreateCollisionMesh for : (nerodia)
CreateCollisionMesh for : (pumpkinseed)
CreateCollisionMesh for : (sinonatrix)
CreateCollisionMesh for : (storeria)
CreateCollisionMesh for : (varada)
CreateCollisionMesh for : (vatakara)
CreateCollisionMesh for : (venturestar)
CreateCollisionMesh for : (wave)
CreateCollisionMesh for : (xylophis)
Creating new galaxy generator 'legacy' version 1
Clearing and re-using previous Galaxy object
StarSystemCache: misses: 0, slave hits: 0, master hits: 0
SectorCache: misses: 101, slave hits: 0, master hits: 1
Sol:
    height fractal: Ellipsoid
    colour fractal: StarG
    seed: 2514954181
Mercury:
    height fractal: MountainsCraters2
    colour fractal: Rock
    seed: 6
Venus:
    height fractal: Barren Rock
    colour fractal: Desert
    seed: 3673958391
Earth:
    height fractal: Mapped
    colour fractal: EarthLikeHeightmapped
    seed: 228984367
Warning: Lua custom Systems definition: Surface starport has been automatically relocated. This is in order to place it on flatter ground to reduce the chance of landing pads being buried. This is not an error as such and you may attempt to move the starport to another location by changing latitude and longitude fields.
      Surface starport name: Brasilia, Body name: Earth, In sector: x = 0, y = 0, z = 0.
Moon:
    height fractal: Mapped2
    colour fractal: Rock2
    seed: 4294967291
Mars:
    height fractal: Mapped
    colour fractal: EarthLikeHeightmapped
    seed: 2977652237
Warning: Lua custom Systems definition: Surface starport has been automatically relocated. This is in order to place it on flatter ground to reduce the chance of landing pads being buried. This is not an error as such and you may attempt to move the starport to another location by changing latitude and longitude fields.
      Surface starport name: Cydonia, Body name: Mars, In sector: x = 0, y = 0, z = 0.
Phobos:
    height fractal: Asteroid4
    colour fractal: BandedRock
    seed: 439771126
Deimos:
    height fractal: Asteroid4
    colour fractal: BandedRock
    seed: 439771126
Warning: Lua custom Systems definition: Surface starport has been automatically relocated. This is in order to place it on flatter ground to reduce the chance of landing pads being buried. This is not an error as such and you may attempt to move the starport to another location by changing latitude and longitude fields.
      Surface starport name: Tomm's Sanctuary, Body name: Deimos, In sector: x = 0, y = 0, z = 0.
Metis:
    height fractal: Asteroid2
    colour fractal: Asteroid
    seed: 4294967198
Adrastea:
    height fractal: Asteroid3
    colour fractal: Asteroid
    seed: 4294963315
Amalthea:
    height fractal: Asteroid2
    colour fractal: BandedRock
    seed: 4294957314
Thebe:
    height fractal: Asteroid
    colour fractal: Rock
    seed: 4293977314
Warning: Lua custom Systems definition: Surface starport has been automatically relocated. This is in order to place it on flatter ground to reduce the chance of landing pads being buried. This is not an error as such and you may attempt to move the starport to another location by changing latitude and longitude fields.
      Surface starport name: Thebe Gas Refinery, Body name: Thebe, In sector: x = 0, y = 0, z = 0.
Io:
    height fractal: MountainsCraters2
    colour fractal: Rock
    seed: 4290011317
Europa:
    height fractal: Barren Rock 2
    colour fractal: Ice
    seed: 2102431459
Ganymede:
    height fractal: Barren Rock
    colour fractal: Rock
    seed: 3062955636
Callisto:
    height fractal: Barren Rock 3
    colour fractal: Rock
    seed: 1272712740
Themisto:
    height fractal: Asteroid3
    colour fractal: Rock
    seed: 134102334
Leda:
    height fractal: Asteroid
    colour fractal: Asteroid
    seed: 4211482628
Himalia:
    height fractal: Asteroid4
    colour fractal: BandedRock
    seed: 1344978
Lysithea:
    height fractal: Asteroid2
    colour fractal: Asteroid
    seed: 3934
Elara:
    height fractal: Asteroid4
    colour fractal: BandedRock
    seed: 128860219
Aega:
    height fractal: Asteroid3
    colour fractal: Asteroid
    seed: 6953
Dione:
    height fractal: Barren Rock 3
    colour fractal: Rock
    seed: 3732948941
Rhea:
    height fractal: RuggedDesert
    colour fractal: Rock
    seed: 584296043
Titan:
    height fractal: MountainsRiversVolcano
    colour fractal: Desert
    seed: 0
Iapetus:
    height fractal: HillsCraters2
    colour fractal: Rock
    seed: 3650944653
Ariel:
    height fractal: Barren Rock 3
    colour fractal: Rock
    seed: 4112814609
Umbriel:
    height fractal: MountainsCraters2
    colour fractal: Rock
    seed: 3404190344
Titania:
    height fractal: Barren Rock 2
    colour fractal: Rock
    seed: 337720919
Oberon:
    height fractal: Barren Rock 3
    colour fractal: Rock
    seed: 2681593006
Proteus:
    height fractal: Asteroid4
    colour fractal: BandedRock
    seed: 1251043226
Triton:
    height fractal: MountainsCraters2
    colour fractal: Rock
    seed: 2173466496
Nereid:
    height fractal: Asteroid3
    colour fractal: Rock
    seed: 285321926
Pluto:
    height fractal: Barren Rock 3
    colour fractal: Rock
    seed: 3144510297
Charon:
    height fractal: MountainsCraters2
    colour fractal: Rock
    seed: 4143618060
CreateCollisionMesh for : (kanara)
CreateCollisionMesh for : (pumpkinseed_police)
error: [string "[T] @pigui/pigui.lua"]:270: not the same number of texts, colors and fonts
stack traceback:
	[C]: in function 'assert'
	[string "[T] @pigui/pigui.lua"]:270: in function 'addFancyText'
	[string "[T] @pigui/game.lua"]:380: in function 'displayReticule'
	[string "[T] @pigui/game.lua"]:489: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:33: in function 'window'
	[string "[T] @pigui/game.lua"]:483: in function 'fun'
	[string "[T] @pigui/pigui.lua"]:53: in function 'withStyleColors'
	[string "[T] @pigui/game.lua"]:482: in function 'game'
	[string "[T] @pigui/pigui.lua"]:65: in function <[string "[T] @pigui/pigui.lua"]:63>
pioneer: Sound.cpp:75: long int Sound::OggFileDataStream::tell(): Assertion `m_data && m_cursor' failed.

PING @ecraven

I put the full text, it's long, but when you read it I edit it. ;)

impaktor commented:
@walterar I took the liberty of editing your post, by making it a code block. Click "edit" on your post and you'll see how I did that.

ecraven commented:
> I put the full text, it's long, but when you read it I edit it. ;)

@walterar Thanks for testing! Can you make sure you are running against the correct pioneer data folder (from my branch) with language english? I've moved and added a few translations, so things only work for english right now (unless you add them manually to your language).

Thanks!

ecraven commented:
> A bit of an update to the icons (hooked them up in the HUD lua too), and to the brake-bar overshoot coloring.

Thanks, I'll add these in later tonight.

walterar commented:
I should have guessed, it's an old problem. It works now.

impaktor commented:
@laarmen do you have any opinions, or plans/desires to review this? 

Just so that we establish @ecraven's "to do" list or time plan for possible review/merge?

laarmen commented:
I'm kinda swamped with life at the moment. I'll try to have a look whenever
I have time and inclination, but consider that a bonus, not a show stopper.

I wouldn't have approved @ecraven in the team if I didn't think he was a
competent fellow.

On Tue, Nov 15, 2016 at 5:12 PM Karl F <notifications@github.com> wrote:

> @laarmen <https://github.com/laarmen> do you have any opinions, or
> plans/desires to review this?
>
> Just so that we establish @ecraven <https://github.com/ecraven>'s "to do"
> list or time plan for possible review/merge?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-260685131>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONRbPythUdND_bJipnN1AwhHJNgz0ks5q-dmmgaJpZM4KwLiH>
> .
>

Toshiwoz commented:
I wonder if it is because I compiled a debug version but it was much slower.

fluffyfreak commented:
@Toshiwoz yes debug is **MUCH** slower. Almost unplayable even on high-end machines, for debugging only.

Toshiwoz commented:
![pioneer 2017-01-30 20-05-13-63](https://cloud.githubusercontent.com/assets/21297356/22449619/4b4402ba-e72f-11e6-956d-050468da5a0e.png)

It seems that the reticule is not in first plain, and probably the planets and station's labels need to be of a different color (darker maybe?)

It also happens when the Hyperspace countdown starts.

impaktor commented on src/PiGui.cpp:
```diff
@@ -85,44 +85,68 @@ ImTextureID PiGui::RenderSVG(std::string svgFilename, int width, int height) {
 	ImFont *font = AddFont(name, size);
-
+	
 	return font;
```
```lisp
;;(add-hook 'before-save-hook 'delete-trailing-whitespace)
(add-hook 'before-save-hook 'whitespace-cleanup) 
```

impaktor commented on data/lang/ui-core/ru.json:
```diff
@@ -1049,7 +1049,7 @@
       "description" : "",
-      "message" : "Моделируем эволюцию вселенной: {age} млрд. лет ;-)"
+      "message" : "Моделируем эволюцию вселенной: {age} млрд. лет ;-) 你正在查看的飞船已经售出"
    },
```
This file shouldn't be included in the PR.

ecraven commented:
Status update: I implemented the multiple-ttfs-as-one-font thing, so now translations should work fine in theory. I'd be very happy if someone on windows could test this.

Toshiwoz commented:
![image](https://cloud.githubusercontent.com/assets/21297356/22622635/ad4121f0-eb0d-11e6-8411-bbf1870758e0.png)


Hi, I just finished compiling the imgui-hud branch and I ran the app in italian, it crashes when I try to start a new game (but the loader actually works).
It seems that there are some missing translation tokens, did I miss to download them from somewhere? Should I have just merged this pull with the master branch? Something else I can't think of?
What if, instead of an error message and exit the game we do have that token in english if not available in the selected language?

nozmajner commented:
As far as I know you can't use it with other languages until those tokens are in lace for them. Which is done automatically by Transifex (not the translation part), and happens only after merge.
A fallback to the English language version would be a nicer way to bug out for sure, especially for mod support. Don't we do that already?

ecraven commented:
Unfortunately, we do not fallback to english, and it isn't trivial to do (as only one language is loaded), however I'll look into how hard it would be to implement (though it would only be useful for debugging, everything in master should be solved by transifex).

impaktor commented:
> I'll look into how hard it would be to implement (though it would only be useful for debugging ...)

Indeed. I wouldn't waste my time on it. Far better to just clean up commits, merge it, and then fix bugs that appear in master as they're discovered. 

In general, it is not reasonable to expect devs to both do all development _and_ testing. The last part is what players can do. And Pioneer makes no claim to be stable, so this can be done from master as well, to optimize dev's time spent on actual development.

impaktor commented on src/PiGui.cpp:
```diff
@@ -318,44 +318,45 @@ void PiGui::NewFrame(SDL_Window *window) {
-	{
-		ImGui::GetIO().MouseDrawCursor = false;
-	}
+		{
```
This indentation looks odd?

ecraven commented:
Hello Toshio!

> I thought he wanted someone to test this with another language like I did...

Thanks for trying that!

Unfortunately I haven't updated any of the language files, so you
wouldn't see anything in a language other than english :-/ Didn't
think of that when I requested the test.

I have however fixed the bug that was blocking this, so with some more
testing and refactoring, it might be good to merge ;D

laarmen commented:
Oh. Cool! I'll have a look at the code during my lunch break then.

Le mar. 18 avr. 2017 à 11:16, ecraven <notifications@github.com> a écrit :

> Hello Toshio!
>
> > I thought he wanted someone to test this with another language like I
> did...
>
> Thanks for trying that!
>
> Unfortunately I haven't updated any of the language files, so you
> wouldn't see anything in a language other than english :-/ Didn't
> think of that when I requested the test.
>
> I have however fixed the bug that was blocking this, so with some more
> testing and refactoring, it might be good to merge ;D
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-294741156>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONfn5ds2sn15oH3ee1PAQPNbcKwzyks5rxH9_gaJpZM4KwLiH>
> .
>

ecraven commented:
> Oh. Cool! I'll have a look at the code during my lunch break then.

Hehe, it's still a mess, but if you could build and see whether it works
for you, that'd be great!

As far as I can tell, it's feature-complete, it just needs some
refactoring, a lot of debug messages need to be removed, and history
needs to be rewritten with nice commit messages.

Thanks for any help with testing it!

laarmen commented:
After a 5mn test, the first think that struck me was the confusion between the "normal" square and the tilted square (I believe the two of them were yellow), we could perhaps start using other shapes ? :-)

I hadn't read the PR description beforehand, so a bit too much of my time has been spent on the reticule-less screens. Oh well.
It is a bit overwhelming when first starting up, there is a LOT of info, but I remember having the exact same feeling when first launching E:D so that can't be such a bad thing, can it? ;-)

On the performance side, sitting on the launchpad at the Earth start, 2 seconds out, I'm at the 60FPS mark. On the other hand, I'm using a laptop with last gen Intel CPU and its attached GPU, so I'm unlikely to notice anything smaller than a 50% drop...

nozmajner commented:
@laarmen I already started making [documentation](https://pioneerwiki.com/wiki/Flight_UI) for it. (not yet linked from the front page) I think it will be quite straightforward after a bit of time using it. (I'm already missing it when I'm not playing on that branch for testing, etc.).

What do you mean by tilted quare? The diamond?

impaktor commented:
I've tested it on my main machine (OpenGL 2.1). At master I have 7-11 fps sitting on the pad in London. With imgui it's 6-7 fps, so pretty much same. In Barnards I have about 45 fps on imgui branch, and I don't know how much on master. (Takes 30 minutes to recompile).

So I don't think imgui will make a previous playable pioneer unplayable. At least not using my machine as reference.

When starting the game, I get a (non-crash triggering) error message:
```
Fonts registered: 2
ImFontAtlas::Build()
imgui: Setting ContainerAtlas
imgui: Setting ContainerAtlas
warning: In function SwapBuffers
On line 324
OpenGL error(s) during frame:
GL_INVALID_OPERATION

GalaxyGenerator::Init()
Creating new galaxy generator 'legacy' version 1
```

Also, I had my nav-tunnel on, apparently; I suggest:
    1. make it yellow, 
    2. make if off by default (perhaps it already is in master?)

And regarding tooltip / mouse over: 
   1. is it possible to require a delay of say ~1.5 s before showing it? Feels instantaneous now. 
   2. when flying (holding in right mouse) the pointer goes away, but will
      still activate the mouse over. I consider that a bug.

ecraven commented:
Fixed bugs, cleaned up history. Note that intermediate steps between the commits probably don't work (or even compile).
![tooltip](https://cloud.githubusercontent.com/assets/215510/25172285/ca9769f4-24f0-11e7-84dd-1a7807d07a14.jpg)
![targets](https://cloud.githubusercontent.com/assets/215510/25172293/cf775ede-24f0-11e7-8322-b030489fb7e4.jpg)
Works for me, tested with Opengl 3.x and 2.1. I'd be happy about any comments on improvements. This is now getting into a state that I'd consider ready to merge.

laarmen commented:
I'll have a read at noon.

Le mer. 19 avr. 2017 à 11:11, ecraven <notifications@github.com> a écrit :

> Fixed bugs, cleaned up history. Note that intermediate steps between the
> commits probably don't work (or even compile).
> [image: tooltip]
> <https://cloud.githubusercontent.com/assets/215510/25172285/ca9769f4-24f0-11e7-84dd-1a7807d07a14.jpg>
> [image: targets]
> <https://cloud.githubusercontent.com/assets/215510/25172293/cf775ede-24f0-11e7-8322-b030489fb7e4.jpg>
> Works for me, tested with Opengl 3.x and 2.1. I'd be happy about any
> comments on improvements. This is now getting into a state that I'd
> consider ready to merge.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-295178404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONdnylKhVVo7uNqzKWbu0ki35ts82ks5rxc_LgaJpZM4KwLiH>
> .
>

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+						 ui.pointOnClock(center, reticuleCircleRadius - 3, 0),
+						 ui.theme.colors.reticuleCircle, 1)
+
+	while true do
```
Now that's just a shame. Why not a for loop ? Much less error-prone and much more readable, too.

ecraven replied:
Thanks for the catch, this is a remnant of an earlier version, rewritten to ```for```

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+		if ratio < 0 then
+			ratio = 0
+		end
+		if ratio > 0 and ratio < 0.001 then
```
The rationale for this ? Might be worth it to put it in a comment, too :)

ecraven replied:
Commented and pushed

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+
+local function displayReticuleDeltaV(center)
+	-- ratio is 1.0 for full, 0.0 for empty
+	local function gauge(ratio, center, radius, color, thickness)
```
Why have this as an inner function ? Also, beware of variable name conflicts.

ecraven replied:
This (and the latter function of the same name) are specialised to left/right, they are not exactly the same. They should be refactored, I just haven't found a very good way to do that yet :-/ I'll probably try to extract much of game.lua into a generic "GUI" toolkit later.

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+	local deltav_current = player:GetCurrentDeltaV()
+	local dvc = deltav_current / deltav_max
+
+	gauge(1.0, center, reticuleCircleRadius + offset, ui.theme.colors.deltaVTotal, thickness)
```
Lots of similar looking code in this section. Perhaps refactor some of it in a variable ? That's more a note for future improvements, though.

ecraven replied:
See above ;)

ecraven replied:
I will address this later, need to think about how to best refactor it.

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+
+end
+
+local brakeNowRatio = 0.90
```
What's that?

ecraven replied:
Commented and tried to explain.

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,512 @@
+local brakeNowRatio = 0.90
+
+local function displayReticuleBrakeGauge(center, ratio, ratio_secondary)
+	local function gauge(ratio, center, radius, color, thickness)
```
Might be worth refactoring with the previous, similarly named inner function.

ecraven replied:
See above

fluffyfreak commented:
Nice, I like the new loading screen addition of the logo.

Just profiling it a little and I see that `PiGui::Render` takes between 9% (GL2.1) and 7% (GL3.x) of the frame time. That appears to be totally inside the Lua `ScopedTable` call so it's all being eaten up by Lua code or the act of calling into it.

laarmen commented:
Just to be clear, is that good or bad in your opinion?

On Thu, Apr 20, 2017 at 12:10 PM Andrew Copland <notifications@github.com>
wrote:

> Nice, I like the new loading screen addition of the logo.
>
> Just profiling it a little and I see that PiGui::Render takes between 9%
> (GL2.1) and 7% (GL3.x) of the frame time. That appears to be totally inside
> the Lua ScopedTable call so it's all being eaten up by Lua code or the
> act of calling into it.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-295665519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONbNpMkwcIP_-6yeABHSLTNHzI3vIks5rxy8hgaJpZM4KwLiH>
> .
>

fluffyfreak commented:
@laarmen I think it shows that there is room for optimisation. It's about equal to the entire game logic for example so it seems a bit high.

impaktor commented on data/lang/core/en.json:
```diff
@@ -223,6 +215,38 @@
       "description" : "",
       "message" : "Communist"
    },
+	 "COMPASS_N" : {
```
a bit weird indentation, compared to the rest of the file.

ecraven replied:
fixed

impaktor commented on data/lang/ui-core/en.json:
```diff
@@ -439,6 +439,66 @@
+			"message" : "Distance to the surface of the target"
+	 },
+   "HUD_DISTANCE_TO_SURFACE_OF_FRAME" : {
+			"description" : "Tooltip: Distance to the surface of the frame of reference.",
```
More of the same indentation complaint in this file.

ecraven replied:
fixed, thanks

impaktor commented on src/Propulsion.h:
```diff
@@ -44,6 +44,7 @@ class Propulsion
 		inline double GetAccelRev() const { return GetThrustRev() / m_dBody->GetMass(); }
 		inline double GetAccelUp() const { return GetThrustUp() / m_dBody->GetMass(); }
 		inline double GetAccelMin() const { return GetThrustMin() / m_dBody->GetMass(); };
+	  inline double GetAccel(Thruster thruster) const { return fabs(m_linThrust[thruster] * m_power_mul / m_dBody->GetMass()); }
```
indentation

ecraven replied:
fixed

nozmajner commented:
@laarmen , @ecraven : How about if the target indicator square looks like this (so it's not at odds with the prograde diamond)? 
![screenshot-20170420-214450](https://cloud.githubusercontent.com/assets/4182678/25249596/b5788a62-2612-11e7-99ae-9efbd5434448.png)

Another request/question: what if the speed bar at the right (that climbs up as you accelerate) would turn red when it approaches the deltaV bar's amount? Like at 80% it would start to turn red and at 90% it's red all the way? So you could have a feel of it from the side of your eye if you are closing to half of your deltaV. So that when you have red bars all over, you are totally screwed. :D
![leftbar](https://cloud.githubusercontent.com/assets/4182678/25250054/6eb23450-2614-11e7-9c05-3686ee85eaf2.png)

laarmen commented:
Yep, much better with the square. And I like the red bar thingy, too.

On Thu, Apr 20, 2017 at 10:02 PM Szlrd <notifications@github.com> wrote:

> @laarmen <https://github.com/laarmen> , @ecraven
> <https://github.com/ecraven> : How about if the target indicator square
> looks like this (so it's not at odds with the prograde diamond)?
> [image: screenshot-20170420-214450]
> <https://cloud.githubusercontent.com/assets/4182678/25249596/b5788a62-2612-11e7-99ae-9efbd5434448.png>
>
> Another request/question: what if the speed bar at the right (that climbs
> up as you accelerate) would turn red when it approaches the deltaV bar's
> amount? Like at 80% it would start to turn red and at 90% it's red all the
> way? So you could have a feel of it from the side of your eye if you are
> closing to half of your deltaV. So that when you have red bars all over,
> you are totally screwed. :D
> [image: leftbar]
> <https://cloud.githubusercontent.com/assets/4182678/25250054/6eb23450-2614-11e7-9c05-3686ee85eaf2.png>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-295884261>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONWHBXA0XrtBK7YubrtgtGVXF2pKuks5rx7nvgaJpZM4KwLiH>
> .
>

nozmajner commented:
These degree indicators inside the reticule feel odd to me. I mean them being inside and that close to the center. Or I'm seeing an in-progress status here? If so, then ignore my nitpicks :)
And their placement is even odder in lower resolutions.
![huddegrees](https://cloud.githubusercontent.com/assets/4182678/25253806/b78c2f6a-2622-11e7-90ab-2c3c5f5d7a07.png)

How about either of these? (both moves the roll a few pixels down, so it doesn't touches the circle.)
This one keeps them inside, but in a better place in my opinion. 
![degree_a](https://cloud.githubusercontent.com/assets/4182678/25254268/00a1bb1e-2625-11e7-8fe6-482cbcb135b1.png)

This one they are on the outside, but less clear, which should be which of the pitch and compass degrees.
![degree_b](https://cloud.githubusercontent.com/assets/4182678/25254125/3ebd67c8-2624-11e7-9692-0dbfb2e89efe.png)

impaktor commented:
> How about if the target indicator square looks like this

I like it, as it is instantly recognizable as a "target".

impaktor commented:
poke-poke @laarmen, will you have time to finish your review of this?

nozmajner commented:
Sorry, forgot to link the target icon: https://www.dropbox.com/s/27u6g8b6e21nlww/icons.svg?dl=0

ecraven commented:
Added the icon to the old commit :)

ecraven commented:
@nozmajner, I'll fix the degree indicators soon, they were not meant to be visible, there's a switch to enable/disable them (for people who don't want them).

impaktor commented:
This needs to be merged A-ZAP. 

In fact, I'm laying down the law now, and announcing that I'll merge this in 12 h, unless someone protestst loudly (@laarmen?), or even better, if someone merges it before me.

I have spoken!

laarmen commented on contrib/imgui/imgui_draw.cpp:
```diff
@@ -119,40 +119,40 @@ static const ImVec4 GNullClipRect(-8192.0f, -8192.0f, +8192.0f, +8192.0f); // La
 
 void ImDrawList::Clear()
 {
-    CmdBuffer.resize(0);
```
Did you just re-indent the imgui upstream file? :-)

That kinda makes the diff hard to read...

ecraven replied:
sorry about this, didn't notice it when commiting, fixed.

laarmen commented on data/libs/Vector.lua:
```diff
@@ -6,9 +6,9 @@ local Vector = {}
 local function binaryOp(a, b, op)
 	local typ = type(b)
 	if(typ == "number") then
-		return Vector(a.x + b, a.y + b, a.z + b)
```
I personally would have split this into its own commit, but this is a really subjective matter :)

ecraven replied:
Once the basics are in, I'll try to work in smaller increments (and thus more logical steps) ;)

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+		else
+			return forward:angle() + adjust
+		end
+  end
```
Indentation malfunction? My internal parser tells me this belongs to angle()

ecraven replied:
fixed

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+
+local reticuleTarget = "frame"
+
+local function displayTargetData(target, center, radius, combatTarget, navTarget, colorLight, colorDark, frameColor, navTargetColor, combatTargetColor)
```
This function is horribly long and undocumented. I'm guessing this displays the informations on the right side of the reticule. But they're not always related to the target (as the player understands the notion), since it can display info related to the current frame, right?

If I'm right, this needs at least a renaming, the current frame isn't a target, it's just a navigation context. AND COMMENTS !

ecraven replied:
should be fixed now

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+	local uiPos = ui.pointOnClock(center, radius, 4.2)
+	local mouse_position = ui.getMousePos()
+	local size = 24
+	if combatTarget or navTarget then
```
See, this is exactly what I'm talking about : from the name of the function, this shouldn't ever register to false, or at least not as late in the function.

ecraven replied:
hope this is clearer now

laarmen commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+	end
+end
+
+local function displayCombatTargetIndicator(center, combatTarget)
```
Since @fluffyfreak seems a little bit concerned about performance, how about saving a function call and only call this method with a valid combat target?

ecraven replied:
I tried to refactor for this

ecraven commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+
+		local uiPos = ui.pointOnClock(center, radius, 2)
+		-- label of target
+		local tooltip = reticuleTarget == "combatTarget" and lui.HUD_CURRENT_COMBAT_TARGET or (reticuleTarget == "navTarget" and lui.HUD_CURRENT_NAV_TARGET or lui.HUD_CURRENT_FRAME)
```
I've used this wherever I would have used the ternary operator in C-likes. Personally, I find it very unreadable, I'll gladly change it to IFs, if this is not idiomatic lua

ecraven commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+		local uiPos = ui.pointOnClock(center, radius, 2)
+		-- label of target
+		local tooltip = reticuleTarget == "combatTarget" and lui.HUD_CURRENT_COMBAT_TARGET or (reticuleTarget == "navTarget" and lui.HUD_CURRENT_NAV_TARGET or lui.HUD_CURRENT_FRAME)
+		ui.addStyledText(uiPos, target.label, colorDark, ui.fonts.pionillium.medium, ui.anchor.left, ui.anchor.baseline, tooltip)
```
done

laarmen commented:
It is idiomatic AFAIK, but this particular case was just too much.

Le dim. 23 avr. 2017 à 22:06, ecraven <notifications@github.com> a écrit :

> *@ecraven* commented on this pull request.
> ------------------------------
>
> In data/pigui/game.lua
> <https://github.com/pioneerspacesim/pioneer/pull/3868#discussion_r112843115>
> :
>
> > +	if showIndicator and (center - position):magnitude() > reticuleCircleRadius * 1.2 then
> +		ui.addIcon(indicator, icon, color, iconSize, ui.anchor.center, ui.anchor.center)
> +	end
> +end
> +
> +local reticuleTarget = "frame"
> +
> +local function displayTargetData(target, center, radius, combatTarget, navTarget, colorLight, colorDark, frameColor, navTargetColor, combatTargetColor)
> +	local player = Game.player
> +	if target then
> +		local velocity = player:GetVelocityRelTo(target)
> +		local position = player:GetPositionRelTo(target)
> +
> +		local uiPos = ui.pointOnClock(center, radius, 2)
> +		-- label of target
> +		local tooltip = reticuleTarget == "combatTarget" and lui.HUD_CURRENT_COMBAT_TARGET or (reticuleTarget == "navTarget" and lui.HUD_CURRENT_NAV_TARGET or lui.HUD_CURRENT_FRAME)
>
> I've used this wherever I would have used the ternary operator in C-likes.
> Personally, I find it very unreadable, I'll gladly change it to IFs, if
> this is not idiomatic lua
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#discussion_r112843115>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONTdUf9mwROiC3VM6ZJru_fZhIedDks5ry69AgaJpZM4KwLiH>
> .
>

ecraven commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+		local approach_speed = position:dot(velocity) / position:magnitude()
+		local speed, speed_unit = ui.Format.Speed(approach_speed)
+
+		ui.addFancyText(uiPos,
```
refactored this as you suggested, makes way more sense than what I had before ;) thank you!

ecraven commented on data/pigui/game.lua:
```diff
@@ -0,0 +1,539 @@
+	local player = Game.player
+	if frame then
+		local frameVelocity = -frame:GetVelocityRelTo(player)
+		local awayFromFrame = player:GetPositionRelTo(frame) * 1.01
```
fixed

ecraven commented:
To my knowledge, I've addressed all comments with the currently committed code, if I overlooked one, please tell me!

nozmajner commented:
The [user documentation for it is here in this not yet linked wiki article. ](https://pioneerwiki.com/wiki/Flight_UI), Maybe put a link into the changelog when this gets merged?

impaktor commented on contrib/imgui/imgui_draw.cpp:
```diff
@@ -1800,15 +1817,31 @@ void ImFont::AddRemapChar(ImWchar dst, ImWchar src, bool overwrite_dst)
+const ImFont::Glyph* ImFont::FindGlyph(unsigned short c, bool report_missing) const
 {
-    if (c < IndexLookup.Size)
+	if (c < IndexLookup.Size)
```
What has happened here is that this contrib file is space indented (4 char wide), but your editor changed this line to tab.

Actually, this file is a bit of a mess. I suspect most is space indented, but stuff you edited is tabs.
 
Doens't get better by github by default showing tabs as 3 chars wide.

I vote: do nothing.

ecraven replied:
thanks for the catch, should be fixed now

impaktor commented on data/lang/core/en.json:
```diff
@@ -1751,6 +1755,30 @@
+   },
+   "UNIT_MILLION_METERS_PER_SECOND" : {
+      "description" : "Speed unit: one million meters per second",
+      "message" : "Mm/s"
```
To me, the 'M' is Mega not million. Same same but different name

impaktor commented on data/lang/core/en.json:
```diff
@@ -1751,6 +1755,30 @@
+      "message" : "km"
+   },
+   "UNIT_MILLION_METERS" : {
+      "description" : "Distance unit: one million meters",
```
same here with mega/million.

impaktor commented:
OK, consensus in IRC (me, ecraven, nozmajer) is that this is finished for merge, but it would be good with a thumbs up from @laarmen first.

Or anyone else: speak up now or forever hold your peace.

nozmajner commented:
![drgenya](https://cloud.githubusercontent.com/assets/4182678/25329955/2942aa36-28de-11e7-882a-c60bec0c1da3.gif)

laarmen commented:
Haven't read the new diffs but I trust ecraven.

Go go go go !

Le lun. 24 avr. 2017 à 11:07, Szlrd <notifications@github.com> a écrit :

> [image: drgenya]
> <https://cloud.githubusercontent.com/assets/4182678/25329955/2942aa36-28de-11e7-882a-c60bec0c1da3.gif>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pioneerspacesim/pioneer/pull/3868#issuecomment-296587733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkONWAlQqUWdCessVaNdNUt4bSk9jEBks5rzGZagaJpZM4KwLiH>
> .
>

fluffyfreak commented:
Yep, go for it.

impaktor commented:
I still get pop-up window saying invalid operation, during start. It doesn't crash though. (on OpenGL 2.1 renderer).

```
Game::CanLoadGame('_exit')
nanosvg: /home/vandelay/usr/src/pioneer/data/icons/icons.svg 1024x1024
nanosvg: /home/vandelay/usr/src/pioneer/data/icons/icons.svg 1024x1024
nanosvg: /home/vandelay/usr/src/pioneer/data/icons/logo.svg 512x512
warning: In function SwapBuffers
On line 324
OpenGL error(s) during frame:
GL_INVALID_OPERATION

In function SwapBuffers
On line 324
OpenGL error(s) during frame:
GL_INVALID_OPERATION
GalaxyGenerator::Init()
Creating new galaxy generator 'legacy' version 1
```
